_posts/2020-06-01-spring-boot-hazelcast.md

```diff
@@ -0,0 +1,232 @@
+
+The members build a separate cluster, and the clients access the cluster from outside.
+To build a cluster we can create a java application, that sets up a `Hazelcast` member,
+but for this example, we can use a prepared `Hazelcast` [server](https://hazelcast.org/imdg/download/#hazelcast-imdg).
+Alternatively, we can start a [docker container](https://hub.docker.com/r/hazelcast/hazelcast)
+as a cluster member. Every server or every docker container will start a new member of the cluster with
+the default configuration.
+
+Now we need to create a client to access the cache cluster. `Hazelcast` uses TCP socket communication. That's
+why it is possible to create a client not only with java. `Hazelcast` provide a list of client written in other languages.
+To keep it simple let's look at how to create a client with Spring.
+
+First, we have to add a `Hazelcast` dependency for the client.
+
+````groovy
+compile group: 'com.hazelcast', name: 'hazelcast-client', version: '3.12.7'
+````
+
+Now we create a `Hazelcast` client in a Spring application.
+
+````java
+@Component
+public class CacheClient {
+
+    private static final String CARS = "cars";
+
+    private HazelcastInstance client = HazelcastClient.newHazelcastClient();
+
+    public Car put(String key, Car car){
+        IMap<String, Car> map = client.getMap(CARS);
+        return map.putIfAbsent(key, car);
+    }
+
+    public Car get(String key){
+        IMap<String, Car> map = client.getMap(CARS);
+        return map.get(key);
+    }
+    
+    // other methods omitted
+
+}
+````
+To create a `Hazelcast` client we need to call the method  `HazelcastClient.newHazelcastClient()` only.
+`Hazelcast` will find the cache cluster automatically. After that, we can use the cache by using the `Map`
+again. If we put or get data from the Map, the `Hazelcast` client connects the cluster to access data.
+
+**Now we can deploy and scale the application and the cache cluster independently.** We can have for example 50
+instances of the application and 5 members of the cache cluster. This the biggest advantage of this topology.
+
+If we have some problems with the cluster, it is easier to identify and to fix this issue, since the
+clients and the cache are separated and not mixed.
+
+This approach has drawbacks. First, whenever we write or read the data from the cluster we need
+net communication. It can take longer than in the approach with the embedded cache. 
+This difference is especially significant by reading access. Second, we have to take care of the
+version compatibility between the cluster members and the clients. 
+
+**The client-server topology should be used, when the deployment of the application is bigger than
+the cluster cache.**
+
+Since we have in the application the clients of the cluster only, we have to find the way to test it.
+It can be done very easily by using [Tescontainers](/spring-boot-flyway-testcontainers/).
```

Comment:
Instead of linking to our article about Testcontainers and Flyway (which doesn't really match the topic), can you link to a Hazelcast module for Testcontainers?

Reply:
There is no Hazelcast module for Testcontainers. I used the GenericContaner in the test.
If it doesn't match the topic, I have no problem to delete this sentence.